util/chunk/codec.go

```diff
@@ -0,0 +1,186 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+import (
+	"encoding/binary"
+	"reflect"
+	"unsafe"
+
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// Codec is used to:
+// 1. encode a Chunk to a byte slice.
+// 2. decode a Chunk from a byte slice.
+type Codec struct {
+	// colTypes is used to check whether a column is fixed sized and what the
+	// fixed size for every element.
+	// NOTE: It's only used for decoding.
+	colTypes []*types.FieldType
+}
+
+// NewCodec creates a new Codec object for encode or decode a Chunk.
+func NewCodec(colTypes []*types.FieldType) *Codec {
+	return &Codec{colTypes}
+}
+
+// Encode encodes a Chunk to a byte slice.
+func (c *Codec) Encode(chk *Chunk) []byte {
+	buffer := make([]byte, 0, chk.MemoryUsage())
+	for _, col := range chk.columns {
+		buffer = c.encodeColumn(buffer, col)
+	}
+	return buffer
+}
+
+func (c *Codec) encodeColumn(buffer []byte, col *column) []byte {
+	var lenBuffer [4]byte
+	// encode length.
+	binary.LittleEndian.PutUint32(lenBuffer[:], uint32(col.length))
+	buffer = append(buffer, lenBuffer[:4]...)
+
+	// encode nullCount.
+	binary.LittleEndian.PutUint32(lenBuffer[:], uint32(col.nullCount))
+	buffer = append(buffer, lenBuffer[:4]...)
+
+	// encode nullBitmap.
+	if col.nullCount > 0 {
+		numNullBitmapBytes := (col.length + 7) / 8
+		buffer = append(buffer, col.nullBitmap[:numNullBitmapBytes]...)
+	}
+
+	// encode offsets.
+	if !col.isFixed() {
+		numOffsetBytes := (col.length + 1) * 4
+		offsetBytes := c.i32SliceToBytes(col.offsets)
+		buffer = append(buffer, offsetBytes[:numOffsetBytes]...)
+	}
+
+	// encode data.
+	buffer = append(buffer, col.data...)
+	return buffer
+}
+
+func (c *Codec) i32SliceToBytes(i32s []int32) (b []byte) {
+	if len(i32s) == 0 {
+		return nil
+	}
+	hdr := (*reflect.SliceHeader)(unsafe.Pointer(&b))
+	hdr.Len = len(i32s) * 4
+	hdr.Cap = hdr.Len
+	hdr.Data = uintptr(unsafe.Pointer(&i32s[0]))
+	return b
+}
+
+// Decode decodes a Chunk from a byte slice, return the remained unused bytes.
+func (c *Codec) Decode(buffer []byte) (*Chunk, []byte) {
+	chk := &Chunk{}
+	for ordinal := 0; len(buffer) > 0; ordinal++ {
+		col := &column{}
+		buffer = c.decodeColumn(buffer, col, ordinal)
+		chk.columns = append(chk.columns, col)
+	}
+	return chk, buffer
+}
+
+// DecodeToChunk decodes a Chunk from a byte slice, return the remained unused bytes.
+func (c *Codec) DecodeToChunk(buffer []byte, chk *Chunk) (remained []byte) {
+	for i := 0; i < len(chk.columns); i++ {
+		buffer = c.decodeColumn(buffer, chk.columns[i], i)
+	}
+	return buffer
```

Comment:
How to use returned `the remained unused bytes`? maybe take a test or example to show how to handle len(remained) > 0?